### PR TITLE
Add JSCS task to the Gulp file

### PR DESCRIPTION
**Pull request by felixzapata**
This pull request is related with the issue #270. 

I 've added the JSCS task to the Gulp file. Instead there is a ``lint`` that runs jshint and jscs. 

The idea of merge the styles is from the task [gulp-jscs-stylish](https://github.com/gonsfx/gulp-jscs-stylish).

A problem with the current JSCS task is that it has not an ``extract`` option like JSHint. This option permits to extract embebed script code from a file. 

So, first I need to check the extension of the file. If the file has HTML extension, I use ``gulp-html-extract`` to obtain inline code.  But, this utility causes another problem. The extracted code can have empty lines at the end of the file. Depending of our ``jscsrc`` file the task, can throw an error.

One posible solution to this, it can be to use ``gulp-trim`` like this:

```javascript
// JSCS has not yet a extract option
    .pipe(gulpif('*.html', extract(), trim()))
    .pipe($.jscs()).on('error', noop)
    .pipe(stylish.combineWithHintResults())
```

But for now, I've disabled the ``disallowTrailingWhitespace`` attribute of the Google ``preset`` rules. Another rule that I've modifed is ``maximumLineLength``.

### CODE REVIEW

**Comment by samccone** [comment]
wow this is so awesome, I am reallllly happy with this

**Comment by samccone** [comment]
@felixzapata we should tweak the CI process to run this lint on every PR as well

File: app/elements/routing.html [hunk 1]
```diff
@@ -7,39 +7,38 @@
...
+    app.route = 'user-info';
+    app.params = data.params;
+    app.scrollPageToTop();
+  });
 
-    // add #! before urls
-    page({
-      hashbang: true
-    });
+  page('/contact', function() {
```
Comment (arthurvr) [inline app/elements/routing.html]: So in the gulpfile we're using `function () {` and here we don't do that space?

File: app/elements/my-greeting/my-greeting.html [hunk 2]
```diff
@@ -7,44 +7,41 @@
 Code distributed by Google as part of the polymer project is also
 subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
 -->
-
 <dom-module id="my-greeting">
   <template>
     <style>
-      :host {
```
Comment (arthurvr) [inline app/elements/my-greeting/my-greeting.html]: Why did you change this?

File: app/elements/my-list/my-list.html [hunk 3]
```diff
@@ -21,29 +20,27 @@
...
-        ready: function() {
-          this.items = [
-            'Responsive Web App boilerplate',
-            'Iron Elements and Paper Elements',
-            'End-to-end Build Tooling (including Vulcanize)',
-            'Unit testing with Web Component Tester',
-            'Routing with Page.js',
-            'Offline support with the Platinum Service Worker Elements'
-          ];
+  (function() {
```
Comment (arthurvr) [inline app/elements/my-list/my-list.html]: Why did you change the indentation here?

File: gulpfile.js [hunk 4]
```diff
@@ -24,6 +24,11 @@ var historyApiFallback = require('connect-history-api-fallback');
 var packageJson = require('./package.json');
 var crypto = require('crypto');
 var polybuild = require('polybuild');
+var stylish = require('gulp-jscs-stylish');
+var extract = require('gulp-html-extract');
+var stylish = require('gulp-jscs-stylish');
+var gulpif = require('gulp-if');
```
Comment (arthurvr) [inline gulpfile.js]: `var gulpIf` for consistency.

File: package.json [hunk 5]
```diff
@@ -11,8 +11,11 @@
     "gulp-changed": "^1.0.0",
     "gulp-cssmin": "^0.1.7",
     "gulp-flatten": "0.0.4",
-    "gulp-if": "^1.2.1",
+    "gulp-html-extract": "0.0.1",
+    "gulp-if": "~1.2.5",
     "gulp-imagemin": "^2.2.1",
+    "gulp-jscs": "~2.0.0",
+    "gulp-jscs-stylish": "~1.1.2",
```
Comment (arthurvr) [inline package.json]: Use carets.

**Comment by arthurvr** [comment]
Thanks @felixzapata! I don't really like this as it's somewhat a hack; but it's better than nothing. Also, is there a way this can work with the JSCS `--fix` feature?

**Comment by felixzapata** [comment]
@arthurevans to use --fix feature you must set the dist folder. You can see an example [here](https://github.com/jscs-dev/gulp-jscs). For this pull request I have not added this option because you have to read every file. Maybe this can be add later. A (minor) problem I have is that in my Sublime Text, the JSCS formatter does not work.

@samccone the deprecated attributes are those that in the [JSHint documentation](http://jshint.com/docs/options/) says that you should use JSCS. You can see this attributes if you search ``Warning``. (For example, ``camelcase``)

File: app/elements/my-list/my-list.html (hunk 3 above)
Comment (felixzapata) [inline app/elements/my-list/my-list.html]: when I run the jscs it throws indentantion warnings messages. I use the prettify Sublime Text plugin to change first of all, the format of: my-list.html, my-greeting.html, routing.html and app.js.

Comment (arthurvr) [inline app/elements/my-list/my-list.html]: This is why those extract plugins aren't really ideal, they take everything inside of the tag, so that results in indentation errors. We do want to stay indenting inside of `script` and `style` tags though. Any possible solution other than disabling the rule?

Comment (felixzapata) [inline app/elements/my-list/my-list.html]: the only problem I have with the extract code, is the extra line at the end of the file. Once I change the indentation using the prettify plugin, I haven´t had more problems with that.

For now, the only solution I think to mantein the rule, is try to use the gulp-trim or any other task to parse the code before passing to the jscs.

Comment (arthurvr) [inline app/elements/my-list/my-list.html]: > For now, the only solution I think to mantein the rule, is try to use the gulp-trim or any other task to parse the code before passing to the jscs.

That sounds like complicated and may end up messy. I'm concerned that people need to maintain and change the gulpfile in their projects afterwards. @addyosmani WDYT?

**Comment by felixzapata** [comment]
@samccone is there any problem with the CI? I've uploaded an small changes but now some tests fail, or throws a timeout.

File: app/elements/my-list/my-list.html (hunk 3 above)
Comment (addyosmani) [inline app/elements/my-list/my-list.html]: Over in #333 we're relying on Crisper to an extent for parts of our script extraction before performing additional operations on it (`.pipe($.if('*.html', $.crisper())) // Extract JS from .html files`). To the best of my knowledge it doesn't clobber indendation and may be worth exploring as an alternative here.

**Comment by addyosmani** [comment]
@felixzapata The Polymer CI on all repos in this org has been a little flaky of late. I would ignore it and we can manually verify changes as needed. Atm, this PR needs to be rebased against master. 

I think we should tackle this in a few steps. I'd like to get proper linting setup and we can worry about `--fix` support in a future release. My current concerns share those of @arthurvr - I think we're close but should look at whether there's better tooling we can swap in here that tackles extraction better.

cc @nicolasgarnier who might have opinions here that can help us and @samccone again as an FYI we're talking about extraction. I'm currently thinking this work could be targeted for 1.2 (1.1 is the next release), but let's see how this goes.

File: app/elements/my-list/my-list.html (hunk 3 above)
Comment (felixzapata) [inline app/elements/my-list/my-list.html]: @addyosmani humm, with Crisper the task throws errors in the html files, similars when I was not using the extract task:

```javascript
var crisper = require('gulp-crisper');
....
.pipe(gulpIf('*.html', crisper()))
```

```
events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: Unexpected token ILLEGAL at elements.html :
     1 |<!--
     2 |@license
---------^
     3 |Copyright (c) 2015 The Polymer Project Authors. All rights reserved.
     4 |This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt

```

Comment (addyosmani) [inline app/elements/my-list/my-list.html]: @nicolasgarnier did you run into this at all with your recipe?

File: gulpfile.js [hunk 6]
```diff
@@ -68,6 +72,10 @@ gulp.task('jshint', function () {
     .pipe(reload({stream: true, once: true}))
     .pipe($.jshint.extract()) // Extract JS from .html files
     .pipe($.jshint())
+    // JSCS has not yet a extract option
+    .pipe(gulpIf('*.html', extract()))
```
Comment (nicolasgarnier) [inline gulpfile.js]: Why not `$.if()` here instead of `gulpIf()`?

File: app/elements/my-list/my-list.html (hunk 3 above)
Comment (nicolasgarnier) [inline app/elements/my-list/my-list.html]: I actually got this when trying to run a function that was expecting JS code on HTML. One differenc between crisper and html-extract is that gulp also outputs the HTML (it separates the JS from the HTML). You can try adding an $if filter to the JSCS function:

Change:
`.pipe($.jscs()).on('error', noop)`
into:
`.pipe($if('*.js, $.jscs())).on('error', noop)`

File: gulpfile.js [hunk 7]
```diff
@@ -24,6 +24,10 @@ var historyApiFallback = require('connect-history-api-fallback');
 var packageJson = require('./package.json');
 var crypto = require('crypto');
 var polybuild = require('polybuild');
+var extract = require('gulp-html-extract');
+var stylish = require('gulp-jscs-stylish');
+var gulpIf = require('gulp-if');
```
Comment (nicolasgarnier) [inline gulpfile.js]: For the 3 above, since they are gulp plugins is it possible to use `gulp-load-plugins`'s `$` instead ?

File: gulpfile.js [hunk 8]
```diff
@@ -68,6 +72,10 @@ gulp.task('jshint', function () {
     .pipe(reload({stream: true, once: true}))
     .pipe($.jshint.extract()) // Extract JS from .html files
     .pipe($.jshint())
+    // JSCS has not yet a extract option
+    .pipe(gulpIf('*.html', extract()))
+    .pipe($.jscs()).on('error', noop)
+    .pipe(stylish.combineWithHintResults())
```
Comment (nicolasgarnier) [inline gulpfile.js]: You could also try `$.jscsStylish.combineWithHintResults()` maybe that works :)

File: gulpfile.js (hunk 6 above)
Comment (arthurvr) [inline gulpfile.js]: :+1:

File: gulpfile.js (hunk 7 above)
Comment (felixzapata) [inline gulpfile.js]: I think that in my first attempt I used the ``$`` but due to other projects I know that some gulp tasks does not work with the ``gulp-load-plugins``. It depends how these are written. 

I will try again to see if I made a mistake with these.

File: app/elements/my-list/my-list.html (hunk 3 above)
Comment (felixzapata) [inline app/elements/my-list/my-list.html]: I need to process the javaScript code inside the HTML files too. That is the reason I added a condition to pass the ``gulp-extract-html``only when the file has the HTML extension.

File: gulpfile.js (hunk 8 above)
Comment (felixzapata) [inline gulpfile.js]: ok, done. It was my fault. I´ve updated the pull request.

File: app/elements/my-list/my-list.html (hunk 3 above)
Comment (nicolasgarnier) [inline app/elements/my-list/my-list.html]: Sorry if I wasn't clear but when using Cripser instead of html-extract you need to filter again on the .js files only becasue Crisper generates both html and js files. For instacne can you try replacing this:

```javascript
...
.pipe($.if('*.html', $.htmlExtract()))
.pipe($.jscs()).on('error', noop)
.pipe($.jscsStylish.combineWithHintResults())
.pipe($.jshint.reporter('jshint-stylish'))
.pipe($.if(!browserSync.active, $.jshint.reporter('fail')));
```

With:


```javascript
...
.pipe($.if('*.html', $.crisper())) // This outputs both JS and HTML files
.pipe($.if('*.js',$.jscs())).on('error', noop) // Need to filter on JS files only because JSCS errors with HTML files
.pipe($.if('*.js',$.jscsStylish.combineWithHintResults()))
.pipe($.jshint.reporter('jshint-stylish'))
.pipe($.if(!browserSync.active, $.jshint.reporter('fail')));
```

Comment (felixzapata) [inline app/elements/my-list/my-list.html]: I've updated the pull request using Crisper but I do not like that the console throws errors with files that you haven't:

```
/github/polymer-starter-kit/app/elements/routing.js
  line 32  col 3  Missing line feed at file end

  ⚠  1 warning

```
Really, the error is related with to routing.html. And, now if the rule ``requireLineFeedAtFileEnd`` is activated, always there will be this errors due to the extract task I think.

Comment (nicolasgarnier) [inline app/elements/my-list/my-list.html]: Mhh yes that's right and it's confusing I agree :/
Then we are back to using html-extract and there is still the indentation issue :D
Maybe our best bet is to file issues to get html-extract handle de-indentation or contribute ourselves to it: https://github.com/FormidableLabs/gulp-html-extract

Comment (felixzapata) [inline app/elements/my-list/my-list.html]: I think the final indentation depends of the user. I mean, you set a style code rules and the developer has to mantein that rules with his editor. 

```gulp-html-extract```uses jsdom to parse a stream from the file. I do not know if it will possible to mantein the indentation.

Maybe another "trick" is to use ```gulp-jsbeautifier``` after the ```extract``` task.

For example (a draft):

```javascript
// JSCS has not yet a extract option
    .pipe($.if('*.html', $.htmlExtract(), $.jsbeautifier({js: {"indent_char": " ", "indent_size": 2}})))
    .pipe($.jscs()).on('error', noop)
    .pipe($.jscsStylish.combineWithHintResults())
```
The problem with this solution is that you have to set the same indentations rules that in your jscs configuration.

Comment (nicolasgarnier) [inline app/elements/my-list/my-list.html]: Another question: do we loose the line number when using `gulp-html-extract` as well? Or does it keeps blank lines (of the suppressed HTML code)?

On possible way would be to simply wait for this PR to land: https://github.com/jscs-dev/node-jscs/pull/1038. Someone is working on adding the `--extract` feature to JSCS...
That would probably give the best results.

Comment (nicolasgarnier) [inline app/elements/my-list/my-list.html]: The JSCS team seems open to include the `extract` feature in their next release. If they confirm this I guess it's a better to wait for this. See https://github.com/jscs-dev/node-jscs/pull/1038

Comment (arthurvr) [inline app/elements/my-list/my-list.html]: [Yeah, I think so too.](https://github.com/PolymerElements/polymer-starter-kit/issues/270#issuecomment-134635316)

Comment (felixzapata) [inline app/elements/my-list/my-list.html]: The line numbers you obtain from the ``gulp-html-extract`` starts counting with the first line of code extracted. I mean, if your script starts in the HTML file at line 14, and the extract throws an error in line 1, really the line is number 14.

I saw the PR in the ``node-jscs``, but in my current project  I could not wait for it so I decided to add this solution and send this PR.

**Comment by felixzapata** [comment]
I think that `gulp-html-extract` does not work in Windows due to the `jsdom` package.

**Comment by felixzapata** [comment]
Hi, only to update the information about my previous comment... [The pull request has been accepted](https://github.com/FormidableLabs/gulp-html-extract/pull/3).

**Comment by addyosmani** [comment]
Can we squash the commits here? Thanks for the update about the upstream PR!

**Comment by felixzapata** [comment]
Hi, yes you can squash the commits. Do you want to update the PR to resolve the conflicts?

**Comment by samccone** [comment]
hi @felixzapata a squash and rebase would be awesome :) Thanks a million :sparkles:

**Comment by felixzapata** [comment]
hi, I will make the rebase as soon as possible.

**Comment by felixzapata** [comment]
Hi, I have discovered that I need to make a fix in the `gulp-html-extract` task to parse better the comments (the license comments at the top of the files). Sorry for that... I am working on it.

**Comment by addyosmani** [comment]
Thanks for the update on where you're at with the patch, @felixzapata :star:

**Comment by felixzapata** [comment]
You're welcome. I have sent [a new pull request](https://github.com/FormidableLabs/gulp-html-extract/pull/5) this morning. 

By the way, the bug is more related with elements without childs instead of the comments.

**Comment by samccone** [comment]
looks to be all merged now @felixzapata :tada: 

awesome work on this PS ;)

**Comment by felixzapata** [comment]
Hi, I´ve updated the pull request and squash some commits (I hope I did it well). Only two notes:
* the CI tests fails and I can´t see the details
* the new tasks related with `polybuild` inside `gulpfile.js` throws a warning because the line (number 201.. the link to the granular configuration) has a comment with more than 100 characters. In this case, I have not changed this line.

Regards.

**Comment by samccone** [comment]
hey @felixzapata mind rebasing this one for us.

Thanks a ton!!!

**Comment by felixzapata** [comment]
Hi @samccone, I will make another rebase as soon as possible.

Regards.

**Comment by felixzapata** [comment]
Hi @samccone I ´ve just made the pull request but I have again (like in my update of a few days ago) a problem with the CI and I can´t see the details.

**Comment by felixzapata** [comment]
hi @addyosmani, is there any problem with the CI? Finally I've seen the details of the error but it is strange because I've reviewed the tests in my local machine and all the tests passed.

By the way, I've updated again the pull request to add the last two commits from the last days.

**Comment by addyosmani** [comment]
I would personally ignore the output of our CI for now. It's somewhat buggy on this repo, however we're happy to manually check the changes. One thing I would ask is that we squash commits here per http://blog.steveklabnik.com/posts/2012-11-08-how-to-squash-commits-in-a-github-pull-request

**Comment by felixzapata** [comment]
Hi, I did a squash a few days ago. Shall I make another?

**Comment by nicolasgarnier** [comment]
@felixzapata Yes please. Basically after you've squashed your commits we should see only one commit instead of the 22 commits that you can currently see at: https://github.com/PolymerElements/polymer-starter-kit/pull/330/commits 
It makes it easier to maintain the changes history down the road as, this way, we'll have only one commit for this feature which is a lot clearer.

Thanks!

**Comment by felixzapata** [comment]
@nicolasgarnier I've just updated the pull request after squash the commits.

**Comment by addyosmani** [comment]
Thanks @felixzapata! @samccone can we get your final LGTM on this before merging in?

**Comment by addyosmani** [comment]
@felixzapata Possible for us to rebase this against master so that it can land?

**Comment by felixzapata** [comment]
@addyosmani done. I've modified the `app.js` file to fix lint errors too.

**Comment by addyosmani** [comment]
Thanks @felixzapata! One last ask from me - would you mind squashing your commits for a cleaner merge? :candy:

**Comment by felixzapata** [comment]
@addyosmani done. Sorry, I always forget to squash my commits. In this case it was because I did not update my remote squashed branch into my working copy

File: gulpfile.js [hunk 9]
```diff
@@ -36,27 +37,23 @@ var AUTOPREFIXER_BROWSERS = [
...
       return path.join('app', stylesPath, src);
     }))
-    .pipe($.changed(stylesPath, {extension: '.css'}))
+    .pipe($.changed(stylesPath, {
+      extension: '.css'
+    }))
     .pipe($.autoprefixer(AUTOPREFIXER_BROWSERS))
     .pipe(gulp.dest('.tmp/' + stylesPath))
-    .pipe($.cssmin())
+    .pipe($.if('*.css', $.cssmin()))
```
Comment (samccone) [inline gulpfile.js]: Why this check? Could we tweak stylesPath to only be css? or at least add this filter higher up in the chain of operations?

File: gulpfile.js [hunk 10]
```diff
@@ -288,4 +307,6 @@ gulp.task('default', ['clean'], function (cb) {
 require('web-component-tester').gulp.init(gulp);
 
 // Load custom tasks from the `tasks` directory
-try { require('require-dir')('tasks'); } catch (err) {}
+try {
+  require('require-dir')('tasks');
+} catch (err) {}
```
Comment (samccone) [inline gulpfile.js]: perhaps lets throw a log here so it is easier to debug in the future

File: package.json [hunk 11]
```diff
@@ -23,6 +27,7 @@
     "gulp-vulcanize": "^6.0.0",
     "jshint-stylish": "^2.0.0",
     "merge-stream": "^0.1.7",
+    "opn": "^1.0.0",
```
Comment (samccone) [inline package.json]: seems to be unused yes?

Comment (felixzapata) [inline package.json]: Yes, I will remove `opn` from package.json file

File: gulpfile.js (hunk 10 above)
Comment (felixzapata) [inline gulpfile.js]: sorry, I do not understand. Are you talking about the `try...catch`? This part is like in the `gulpfile.js` file in the branch master.

File: gulpfile.js (hunk 9 above)
Comment (felixzapata) [inline gulpfile.js]: Sorry, I think this was due to an incorrect rebase in the past. I will update this task to look like the `gulpfile.js` in the branch master.

File: gulpfile.js (hunk 10 above)
Comment (samccone) [inline gulpfile.js]: Yep I am, your formatting change just made me think about how hard this could be to debug potentially

Comment (felixzapata) [inline gulpfile.js]: ok, do you want to add a `console.log(err)` inside the `catch`?

File: gulpfile.js [hunk 12]
```diff
@@ -88,39 +85,52 @@ var optimizeHtmlTask = function (src, dest) {
...
+      once: true
+    }))
+
+  // Extract JS from .html files
+  .pipe($.jshint.extract())
     .pipe($.jshint())
+
+  // JSCS has not yet a extract option
+  .pipe($.if('*.html', $.htmlExtract()))
+    .pipe($.jscs()).on('error', noop)
```
Comment (samccone) [inline gulpfile.js]: nooping the error kinda stinks right?

Comment (samccone) [inline gulpfile.js]: a console.bell and console.warn seems like a pretty decent solution here instead

Comment (felixzapata) [inline gulpfile.js]: I've just discovered that with version 3.0.0 of `gulp-jscs` you do not need the noop anymore. The problem was that `jscs` task threw an error,  and if you would catched this error and shown it via console, you could get (it depends of the kind of error) duplicate error messages.

I will update the version of `gulp-jscs` in the `package.json` file.

File: gulpfile.js (hunk 10 above)
Comment (samccone) [inline gulpfile.js]: that would be great :)

File: gulpfile.js (hunk 12 above)
Comment (samccone) [inline gulpfile.js]: :+1:

File: gulpfile.js (hunk 10 above)
Comment (felixzapata) [inline gulpfile.js]: ok, but if I add this line, always will show this error:

``
{ [Error: ENOENT: no such file or directory, scandir '/myhomefolder/github/polymer-starter-kit/tasks']
  errno: -2,
  code: 'ENOENT',
  syscall: 'scandir',
  path: '/myhomefolder/github/polymer-starter-kit/tasks' }
``

because the `tasks` folder does not exist in the project. Personally, when I use Gulp tasks, I like to create this folder and separate the tasks into js files. For example:
- lint.js : with all related with the lint
- serve.js: with all related with the server tasks
- etc

I have previous versions of the Polymer Starter Kit with these files to use in some personal projects.

Comment (samccone) [inline gulpfile.js]: hmmm ok fair argument, we can punt on it for now.

**Comment by samccone** [comment]
woot @felixzapata ... I did some cleanup work and such but this just landed :dancer: 

c4c411c11fc669e94247cd55c6c36c0ebe7cbc8c

Thanks a tonnnnnnn :sparkles: :sparkling_heart:
